File: packages/registry/src/negotiate.ts
import { type ComponentNode, SANDBOX_HTML_TYPE, type UISpec } from "@kohaku-ui/spec-core";
import semver from "semver";
import type { ResolvedCatalog } from "./catalog.js";
import type { SurfaceCapabilities } from "./types.js";

export interface Downgrade {
  id: string;
  from: string;
  to: string;
  reason: string;
}

const TERMINAL_FALLBACK = "presentMarkdown";

/**
 * Capability negotiation. Downgrades parts the surface does not implement along the definition-side
 * fallback chain. The terminal is presentMarkdown (a text fallback).
 */
export function negotiate(
  spec: UISpec,
  catalog: ResolvedCatalog,
  surface: SurfaceCapabilities,
): { spec: UISpec; downgrades: Downgrade[] } {
  const downgrades: Downgrade[] = [];
  // Pin the catalog version on downgraded nodes too (a missing version causes inconsistency on re-validation / re-negotiation).
  const ver = (type: string): { version?: string } => {
    const v = catalog.get(type)?.version;
    return v != null ? { version: v } : {};
  };

  const components = spec.components.map((node): ComponentNode => {
    if (node.type === SANDBOX_HTML_TYPE) {
      if ((surface.maxTier ?? "L2") === "L2") return node;
      downgrades.push({
        id: node.id,
        from: node.type,
        to: TERMINAL_FALLBACK,
        reason: `surface does not allow tier L2`,
      });
      return {
        id: node.id,
        type: TERMINAL_FALLBACK,
        ...ver(TERMINAL_FALLBACK),
        props: { markdown: "(Free-form generated components cannot be displayed on this surface)" },
      };
    }

    if (supports(surface, node.type, node.version)) return node;

    // Follow the fallback chain
    let current = node;
    const visited = new Set<string>([node.type]);
    for (;;) {
      const def = catalog.get(current.type);
      const fb = def?.fallback;
      if (fb == null || visited.has(fb.type)) {
        if (current.type !== TERMINAL_FALLBACK) {
          downgrades.push({
            id: node.id,
            from: node.type,
            to: TERMINAL_FALLBACK,
            reason: `surface lacks "${node.type}" and no usable fallback chain`,
          });
          return {
            id: node.id,
            type: TERMINAL_FALLBACK,
            ...ver(TERMINAL_FALLBACK),
            props: { markdown: `(Component ${node.type} is not available on this surface)` },
          };
        }
        return current;
      }
      visited.add(fb.type);
      const next: ComponentNode = {
        id: current.id,
        type: fb.type,
        ...ver(fb.type),
        props: fb.mapProps(current.props),
        // If the downgrade target does not accept children (capabilities.children === "none"), drop
        // the children. Symmetric with the data downgrade check (next line). Leaving children on a
        // children:"none" part would produce CHILDREN_NOT_SUPPORTED in structural validation.
        ...(current.children != null && catalog.get(fb.type)?.capabilities.children !== "none"
          ? { children: current.children }
          : {}),
        ...(current.data != null && catalog.get(fb.type)?.capabilities.data !== "none"
          ? { data: current.data }
          : {}),
      };
      if (supports(surface, fb.type, catalog.get(fb.type)?.version)) {
        downgrades.push({
          id: node.id,
          from: node.type,
          to: fb.type,
          reason: `surface lacks "${current.type}"`,
        });
        return next;
      }
      current = next;
    }
  });

  if (downgrades.length === 0) return { spec, downgrades };

  return {
    spec: {
      ...spec,
      components,
      provenance: {
        ...spec.provenance,
        // If an existing fallback is present (e.g. "generation" from a generation failure), negotiation
        // overwrites it last-writer-wins (a simplification of one downgrade trace per Spec; recording
        // multiple kinds together = turning it into an array is future work).
        fallback: {
          from: downgrades.map((d) => `${d.id}:${d.from}`).join(","),
          reason: "capability negotiation",
          kind: "negotiation",
        },
      },
    },
    downgrades,
  };
}

/**
 * Memoizes the result of semver.satisfies (performance). The key space is bounded, coming from the
 * catalog (part version) × surface (supports range), but to avoid unbounded growth in a long-lived
 * process it is cleared once a maximum count is reached.
 */
const SATISFIES_CACHE_MAX = 1000;
const satisfiesCache = new Map<string, boolean>();

function satisfiesMemo(version: string, range: string): boolean {
  const key = `${version}|${range}`;
  const cached = satisfiesCache.get(key);
  if (cached !== undefined) return cached;
  const result = semver.satisfies(version, range);
  // On reaching the limit, clear everything naively (an LRU is not needed; the key space is bounded and rarely reached).
  if (satisfiesCache.size >= SATISFIES_CACHE_MAX) satisfiesCache.clear();
  satisfiesCache.set(key, result);
  return result;
}

function supports(surface: SurfaceCapabilities, type: string, version: string | undefined): boolean {
  const range = surface.supports[type];
  if (range == null) return false;
  if (version == null) return true;
  return satisfiesMemo(version, range);
}

File: python/kohaku/src/kohaku/registry/negotiate.py
"""Capability negotiation (Port of TS negotiate.ts).

Downgrades parts the surface does not implement along the definition-side fallback chain.
The terminal is presentMarkdown (a text fallback).
"""

from __future__ import annotations

from dataclasses import dataclass

from kohaku.spec import SANDBOX_HTML_TYPE, ComponentNode, ProvenanceFallback, UISpec

from .catalog import ResolvedCatalog
from .semver_util import satisfies
from .types import SurfaceCapabilities

_TERMINAL_FALLBACK = "presentMarkdown"


@dataclass(frozen=True)
class Downgrade:
    id: str
    from_: str
    to: str
    reason: str


@dataclass(frozen=True)
class NegotiateResult:
    spec: UISpec
    downgrades: list[Downgrade]


def _make_node(
    node_id: str, type_: str, version: str | None, props: dict[str, object]
) -> ComponentNode:
    """Construct while omitting the optional field (version) key itself rather than passing None."""
    extra: dict[str, object] = {"version": version} if version is not None else {}
    return ComponentNode(id=node_id, type=type_, props=props, **extra)  # type: ignore[arg-type]


def negotiate(
    spec: UISpec, catalog: ResolvedCatalog, surface: SurfaceCapabilities
) -> NegotiateResult:
    downgrades: list[Downgrade] = []

    def ver(type_: str) -> str | None:
        """Pin the catalog version on downgraded nodes too (a missing version causes inconsistency on re-validation / re-negotiation)."""
        definition = catalog.get(type_)
        return definition.version if definition is not None else None

    def negotiate_node(node: ComponentNode) -> ComponentNode:
        if node.type == SANDBOX_HTML_TYPE:
            if (surface.maxTier or "L2") == "L2":
                return node
            downgrades.append(
                Downgrade(
                    id=node.id,
                    from_=node.type,
                    to=_TERMINAL_FALLBACK,
                    reason="surface does not allow tier L2",
                )
            )
            return _make_node(
                node.id,
                _TERMINAL_FALLBACK,
                ver(_TERMINAL_FALLBACK),
                {"markdown": "(Free-form generated components cannot be displayed on this surface)"},
            )

        if _supports(surface, node.type, node.version):
            return node

        # Follow the fallback chain
        current = node
        visited = {node.type}
        while True:
            definition = catalog.get(current.type)
            fb = definition.fallback if definition is not None else None
            if fb is None or fb.type in visited:
                if current.type != _TERMINAL_FALLBACK:
                    downgrades.append(
                        Downgrade(
                            id=node.id,
                            from_=node.type,
                            to=_TERMINAL_FALLBACK,
                            reason=f'surface lacks "{node.type}" and no usable fallback chain',
                        )
                    )
                    return _make_node(
                        node.id,
                        _TERMINAL_FALLBACK,
                        ver(_TERMINAL_FALLBACK),
                        {"markdown": f"(Component {node.type} is not available on this surface)"},
                    )
                return current
            visited.add(fb.type)
            fb_def = catalog.get(fb.type)
            # Drop children / data if the downgrade target does not accept them (capabilities is "none").
            # Leaving them would produce CHILDREN_NOT_SUPPORTED / DATA_FORBIDDEN in structural validation.
            keep_children = (
                current.children is not None
                and (fb_def is None or fb_def.capabilities.children != "none")
            )
            keep_data = current.data is not None and (
                fb_def is None or fb_def.capabilities.data != "none"
            )
            # Omit the optional-field key itself rather than passing None (coexisting with wire validation's null rejection)
            extra: dict[str, object] = {}
            if ver(fb.type) is not None:
                extra["version"] = ver(fb.type)
            if keep_children:
                extra["children"] = current.children
            if keep_data:
                extra["data"] = current.data
            next_node = ComponentNode(
                id=current.id,
                type=fb.type,
                props=fb.map_props(current.props),
                **extra,  # type: ignore[arg-type]
            )
            if _supports(surface, fb.type, fb_def.version if fb_def is not None else None):
                downgrades.append(
                    Downgrade(
                        id=node.id,
                        from_=node.type,
                        to=fb.type,
                        reason=f'surface lacks "{current.type}"',
                    )
                )
                return next_node
            current = next_node

    components = [negotiate_node(node) for node in spec.components]

    if len(downgrades) == 0:
        return NegotiateResult(spec=spec, downgrades=downgrades)

    # If an existing fallback is present (e.g. "generation" from a generation failure), negotiation
    # overwrites it last-writer-wins (a simplification of one downgrade trace per Spec).
    new_provenance = spec.provenance.model_copy(
        update={
            "fallback": ProvenanceFallback.model_validate(
                {
                    "from": ",".join(f"{d.id}:{d.from_}" for d in downgrades),
                    "reason": "capability negotiation",
                    "kind": "negotiation",
                }
            )
        }
    )
    return NegotiateResult(
        spec=spec.model_copy(update={"components": components, "provenance": new_provenance}),
        downgrades=downgrades,
    )


def _supports(surface: SurfaceCapabilities, type_: str, version: str | None) -> bool:
    version_range = surface.supports.get(type_)
    if version_range is None:
        return False
    if version is None:
        return True
    return satisfies(version, version_range)
